# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [208]:
#lambda >input< x: what we want to do

def add_two(x):
    return x + 2
print(add_two(4))

a_var = 4
print((lambda x: x+2)(a_var))

6
6


#### Saving to a Variable

In [209]:
f_test = lambda x: x+2
f_test(4)

6

#### Multiple Inputs

In [210]:
print((lambda x,y,z: x*y*z)(3,5,8))
x_test = lambda x,y,z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [211]:
def multiply(function, num):
    return function(num)
multiply(lambda x: x*x, 4)
def squared_nums(num):
    return num ** 2
multiply(squared_nums, 4)

16

#### Returning a Lambda from a Function

In [212]:
def multiply_test(num):
    return num * 4

def return_func():
    test = 4
    def multiply(num):
        return num*2
    return multiply

f_return = return_func()
print(f_return(4))

def return_lamb(b,c):
    return lambda x,a: x+a+b+c
r_lamb = return_lamb(4,5)
print(r_lamb)
print(r_lamb(5,5))

8
<function return_lamb.<locals>.<lambda> at 0x00000247B626F2E0>
19


#### If Statements within Lambdas

In [213]:
f_conditional = lambda num: num * 2 if num > 10 else num + 2
print(f_conditional(8))
print(f_conditional(20))

10
40


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [214]:
f = lambda x: x ** 3
result = f(2)  
print(result)  


8


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [215]:
numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]


def squared_normal(val):
    return val ** 2
lst= []
for num in numbers:
    lst.append(squared_normal(num))
    print(lst)

[16]
[16, 121]
[16, 121, 400]
[16, 121, 400, 9]
[16, 121, 400, 9, 225]
[16, 121, 400, 9, 225, 400]
[16, 121, 400, 9, 225, 400, 121]
[16, 121, 400, 9, 225, 400, 121, 4]
[16, 121, 400, 9, 225, 400, 121, 4, 9]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25, 1]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25, 1, 1936]
[16, 121, 400, 9, 225, 400, 121, 4, 9, 36, 25, 1, 1936, 1]


In [216]:
def squared_map(num, num2):
    if num  < 10 and num2 < 10:
        return num**2, num2 **2
    else:
        return num, num2
    
numbers = [4,11,20,3,15,20,11,2,3,6,5,1,44,1]
more_nums = [4,10,3,2,6,10,15,3,4,7,6,0,41,2]

for val1, val2 in zip(numbers, more_nums):
    print(squared_map(val1, val2))

squared_numbers_mapped = list(map(squared_map, numbers, more_nums))
print(squared_numbers_mapped)
another_squared_nums_mapped = list(map(squared_normal, numbers))

(16, 16)
(11, 10)
(20, 3)
(9, 4)
(15, 6)
(20, 10)
(11, 15)
(4, 9)
(9, 16)
(36, 49)
(25, 36)
(1, 0)
(44, 41)
(1, 4)
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6), (20, 10), (11, 15), (4, 9), (9, 16), (36, 49), (25, 36), (1, 0), (44, 41), (1, 4)]


#### Using Lambda's with Map

In [217]:
list(map(lambda x,y: (x**2, y**2)if x < 10 and y < 10 else (x,y), numbers, more_nums))

[(16, 16),
 (11, 10),
 (20, 3),
 (9, 4),
 (15, 6),
 (20, 10),
 (11, 15),
 (4, 9),
 (9, 16),
 (36, 49),
 (25, 36),
 (1, 0),
 (44, 41),
 (1, 4)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [218]:
my_list = [1, 2, 3, 4, 5]
new_list = list(map(lambda x: x*2 - 1, my_list))
print(new_list)

picture = ['', 'abc', 'ded', '']
list(map(lambda x: '*' + x + '*' if len(x) > 1 else '*****', picture))



[1, 3, 5, 7, 9]


['*****', '*abc*', '*ded*', '*****']

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [219]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angela']
def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False
list(filter(a_names, names))

['Andy', 'Angela']

In [235]:
for name in names:
    print(name, name[0], name[1:])

Bob B ob
Andy A ndy
Max M ax
Evan E van
Angela A ngela


#### Using Lambda's with Filter()

In [236]:
list(filter(lambda name: True if name[0].lower() == 'a' else False, names))

['Andy', 'Angela']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [16]:
import statistics
import numpy as np

print(statistics.mean(numbers))
print(np.mean(numbers))
print(list(filter(lambda x: True if x >= sum(numbers)/ len(numbers) else False, numbers)))
print(list(filter(lambda x: True if x >= np.mean(numbers)else False, numbers)))



NameError: name 'numbers' is not defined

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [17]:
def add_nums_breakable(num):
    return num + add_nums_breakable(num-1)
add_nums_breakable(5)

RecursionError: maximum recursion depth exceeded

#### Implementing a Base Case

In [25]:
def add_num(num):
    if num <= 1:
        print('add_nums(1) = 1')
        return num
    else:
        print(f'add_nums({num} = {num} + add_nums({num - 1})')
        return num + add_nums(num-1)

#### Writing a Factorial Function

In [42]:
def factorial(num):
       if num <= 1:
        print(f'factorial({num}) = {num})')
        return 1
       else:
        print(f'factorial({num} = {num} + factorial({num - 1})')
        return num + factorial(num-1)
print(factorial(5))
5*4*3*2*1

factorial(5 = 5 + factorial(4)
factorial(4 = 4 + factorial(3)
factorial(3 = 3 + factorial(2)
factorial(2 = 2 + factorial(1)
factorial(1) = 1)
15


120

#### In-Class Exercise #4 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [98]:
def subtract_numbers(n, y):
    if y == 0:
        return n
    else:
        print(n,y)
        return subtract_numbers(n-1, y-1)
subtract_numbers(25,5)



#def subtract_all_numbers(n):
  #  if n == 0:
  #      return 0
  #  else:
  #      return n - subtract_all_numbers(n - 1)
    
#result = subtract_all_numbers(5)
#print(result)
 



25 5
24 4
23 3
22 2
21 1


20

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [122]:
#print(type(range(0,10)))
def my_range(stop, start=0,step=2):
    while start < stop:
        yield start
        start += step
[i for i in my_range(10)]

[0, 2, 4, 6, 8]

#### In-Class Exercise #5 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [128]:
def squared_nums(n):
    while n > 0:
        yield n ** 2
        n -= 1
[i for i in squared_nums(5)]


[25, 16, 9, 4, 1]

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [179]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
new_list = list(filter(lambda x: x != "  ", places))
print(new_list)


[' ', 'Argentina', ' ', 'San Diego', '', '', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [180]:

author = ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']
author.sort(key=lambda x: x.split()[-1].lower())
print(author)


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [181]:
# F = (9/5)*C + 32
my_list = [('Nashua', 32), ('Boston', 12), ('Los Angeles', 44), ('Miami', 29)]
converted_list = list(map(lambda x: (x[0], (x[1] * 9/5) + 32), my_list))
print(converted_list)


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [183]:
def fib(num):
    if num == 0 or num == 1:
        return 1
    else:
        return fib(num-1) + fib(num-2)

for i in range(6):
    print(f"Iteration {i}: {fib(i)}")


Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8
